main/updater.js

```diff
@@ -1,69 +1,158 @@
-const { BrowserWindow, app, dialog, Notification } = require('electron')
+const { BrowserWindow, app, Notification, shell } = require('electron')
 const { autoUpdater } = require('electron-updater')
-const fs = require('fs')
+const { ipcMain } = require('electron/main')
 const log = require('electron-log').scope('updater')
+const { showDialogSync } = require('./dialog')
+const { ipcMainEvents } = require('./ipc')
 
 // must be global to avoid gc
 let updateNotification = null
 
+let checkingManually = false
+
+function beforeQuitCleanup () {
+  BrowserWindow.getAllWindows().forEach(w => w.removeAllListeners('close'))
+  app.removeAllListeners('window-all-closed')
+}
+
 function setup (/** @type {import('./typings').Context} */ _ctx) {
   autoUpdater.autoDownload = false // we download manually in 'update-available'
 
-  autoUpdater.on('error', err => log.error('error', err))
-  autoUpdater.on('update-available', async (/* { version, releaseNotes } */) => {
-    try {
-      log.info('update-available, downloading..')
-      await autoUpdater.downloadUpdate()
-      log.info('update-available, downloaded')
-    } catch (err) {
-      log.error('update-available error', err)
+  autoUpdater.on('error', onUpdaterError)
+  autoUpdater.on('update-available', onUpdateAvailable)
+  autoUpdater.on('update-not-available', onUpdateNotAvailable)
+  autoUpdater.on('update-downloaded', onUpdateDownloaded)
+
+  // built-in updater != electron-updater
+  // https://github.com/electron-userland/electron-builder/pull/6395
+  require('electron').autoUpdater.on('before-quit-for-update', beforeQuitCleanup)
+}
+
+module.exports = async function (/** @type {import('./typings').Context} */ ctx) {
+  if (['test', 'development'].includes(process.env.NODE_ENV ?? '')) {
+    ctx.manualCheckForUpdates = () => {
+      showDialogSync({
+        title: 'Not available in development',
+        message: 'Yes, you called this function successfully.',
+        type: 'info',
+        buttons: ['Close']
+      })
     }
+    return
+  }
+
+  setup(ctx)
+
+  checkForUpdatesInBackground() // async check on startup
+  setInterval(checkForUpdatesInBackground, 43200000) // every 12 hours
+
+  // enable on-demand check via Tray menu
+  ctx.manualCheckForUpdates = () => {
+    checkingManually = true
+    checkForUpdatesInBackground()
+  }
+}
+
+function checkForUpdatesInBackground () {
+  ipcMain.emit(ipcMainEvents.UPDATE_CHECK_STARTED)
+
+  // TODO: replace this with autoUpdater.checkForUpdatesAndNotify()
+  autoUpdater.checkForUpdates()
```

Comment:
This todo comment is coming from the original source code. I am not sure if it's still relevant. Do you remember @lidel why you would like to switch from `autoUpdater.checkForUpdates()` to `autoUpdater.checkForUpdatesAndNotify()`?

Reply:
I vaguely remember that the rationale was that it is less code to maintain, and that meant smaller surface for breaking autoupdates when upstream APIs change in subtle ways (happened in the past)